src/com/amazon/ion/impl/IonReaderBinaryRawX.java

```diff
@@ -1181,81 +1202,106 @@ protected final Timestamp readTimestamp(int len) throws IOException
         }
     }
 
-    protected final String readString(int len) throws IOException
+    protected final String readString(int numberOfBytes) throws IOException
     {
-        // len is bytes, which is greater than or equal to java
-        // chars even after utf8 to utf16 decoding nonsense
-        // the char array is way faster than using string buffer
-        char[] chars = new char[len];
-        int    c, ii = 0;
-        int    save_limit = NO_LIMIT;
-        if (_local_remaining != NO_LIMIT) {
-            save_limit = _local_remaining - len;
+        // If the string we're reading is small enough to fit in our reusable buffer, we can avoid the overhead
+        // of looping and bounds checking.
+        if (numberOfBytes < utf8InputBuffer.array().length) {
+            return readStringWithReusableBuffer(numberOfBytes);
         }
-        _local_remaining = len;
-        while (!isEOF()) {
-            c = readUnicodeScalar();
-            if (c < 0) throwUnexpectedEOFException();
-            if (c < 0x10000) {
-                chars[ii++] = (char)c;
+
+        // Otherwise, allocate a one-off decoding buffer that's large enough to hold the string
+        // and prepare to decode the string in chunks.
+        CharBuffer decodingBuffer = CharBuffer.allocate(numberOfBytes);
+        utf8CharsetDecoder.reset();
+
+        int save_limit = NO_LIMIT;
+        if (_local_remaining != NO_LIMIT) {
+            save_limit = _local_remaining - numberOfBytes;
+        }
+        _local_remaining = numberOfBytes;
+
+        // The following loop will:
+        // * Fill the input buffer with utf8 bytes
+        // * Write decoded chars to the decoding buffer
+        // * Move any remaining partial character bytes to the front of the buffer
+        // * Repeat until the requested number of bytes have been decoded.
+        // * Create a new String object from the contents of the decoding buffer.
+        int totalBytesRead = 0;
+        int carryoverBytes = 0;
+        while (totalBytesRead < numberOfBytes) {
+
+            int bytesRemaining = numberOfBytes - totalBytesRead;
+            // When decoding, it's possible to have 'carryover' bytes left in the buffer which represent
+            // a partial unicode character. We need to leave these in the buffer so that we can complete the
+            // partial character in the next call to read().
+            int capacityRemaining = utf8InputBuffer.array().length - carryoverBytes;
+            int bytesToRead = Math.min(bytesRemaining, capacityRemaining);
+
+            int bytesRead = read(utf8InputBuffer.array(), carryoverBytes, bytesToRead);
+            if (bytesRead <= 0) {
```

Comment:
So, to confirm, it returns 0 whenever EOF is reached, even if some bytes were successfully read before reaching EOF, correct?

Reply:
It returns `0` to indicate `EOF` rather than -1. Here's the comment in question:

```java
// NB this method does not follow the contract of InputStream.read, it will return 0 at EOF
//    It is unclear what the implication to the rest of the system to make it 'conform'
public final int read(byte[] dst, int offset, int length) throws IOException
{
   //...
```

The JavaDoc for `InputStream#read` states:
```
     * @return     the total number of bytes read into the buffer, or
     *             <code>-1</code> if there is no more data because the end of
     *             the stream has been reached.
```

Reply:
Okay, so when `length` bytes are requested, but EOF is reached when only `n` bytes were read, the method will return with a value of `n` first, and the next time it is called it will return 0 instead of -1, correct?

Reply:
From what I can tell, [it will return 0 as soon as it realizes it's reached EOF](https://github.com/amzn/ion-java/blob/d5939190808b815ec940ef5d45db0d39e9c3fb63/src/com/amazon/ion/impl/UnifiedInputStreamX.java#L365-L384), regardless of whether it has returned all of the available data. I admit I didn't dig into it too deeply seeing as the existing behavior I was preserving was to simply throw an `UnexpectedEOFException` if it ever occurred. I found the comment on `UnifiedInputStreamX#read` to be unsettling, so I mentioned it in the code.